File: src/core/session-loop.ts
import { createInterface } from 'node:readline/promises';
import { stdin, stdout } from 'node:process';
import type { Session, SessionOutcome, Quote } from '../types.js';
import { streamQuestion, getConversationModel, getModelForFile } from '../llm/client.js';
import { getSystemPrompt } from '../llm/prompts.js';
import { validateResponse, repromptIfInvalid, getFallbackQuestion } from '../llm/validator.js';
import { detectFilePaths, readCodeFile, formatCodeContext } from '../storage/file-reader.js';
import { addTurn, addCodeFile } from './session.js';
import { checkAndSummarize, buildMessageArray, estimateTokens } from './context-manager.js';
import { selectQuote } from '../quotes/selector.js';
import {
  streamToTerminal,
  newLine,
  displaySessionHeader,
  displayDim,
} from '../ui/renderer.js';

const EXIT_COMMANDS = new Set(['exit', 'quit']);
const SOLVE_COMMANDS = new Set(['solved', 'found it']);

export interface SessionLoopResult {
  outcome: SessionOutcome;
}

export async function runSessionLoop(session: Session): Promise<SessionLoopResult> {
  const rl = createInterface({ input: stdin, output: stdout });
  const codeFiles = new Map<string, string>();
  let rollingSummary: string | undefined;
  let lifelineOffered = false;

  displaySessionHeader(session.intensity);

  // Read initial code files from problem description
  const initialPaths = detectFilePaths(session.problemDescription);
  for (const filePath of initialPaths) {
    const content = await readCodeFile(filePath);
    if (content) {
      codeFiles.set(filePath, content);
      addCodeFile(session, filePath);
    }
  }

  // Generate opening question
  let model = initialPaths.length > 0 ? getModelForFile(initialPaths[0]) : getConversationModel();
  const codeContext = formatCodeContext(codeFiles);
  let systemPrompt = getSystemPrompt(
    session.intensity,
    session.problemDescription,
    codeContext || undefined,
    rollingSummary,
    1,
  );

  const openingQuestion = await generateValidQuestion(
    systemPrompt,
    [{ role: 'user', content: session.problemDescription }],
    model,
    session.intensity,
  );

  newLine();
  displayDim(`[Turn 1]`);
  console.log(openingQuestion);
  newLine();

  // Add opening turn with empty response (to be filled when user responds)
  let currentQuestion = openingQuestion;

  try {
    while (true) {
      const input = await rl.question('> ');
      const trimmed = input.trim();

      if (!trimmed) continue;

      const lower = trimmed.toLowerCase();

      // Check exit commands
      if (EXIT_COMMANDS.has(lower)) {
        addTurn(session, currentQuestion, '[exited]', model);
        rl.close();
        return { outcome: 'abandoned' };
      }

      // Check solve commands
      if (SOLVE_COMMANDS.has(lower)) {
        addTurn(session, currentQuestion, '[solved]', model);
        rl.close();
        return { outcome: 'solved' };
      }

      // Check "I give up"
      if (lower === 'i give up') {
        if (!lifelineOffered) {
          lifelineOffered = true;
          addTurn(session, currentQuestion, trimmed, model);

          // Generate one lifeline question
          const lifelinePrompt = getSystemPrompt(
            session.intensity,
            session.problemDescription,
            codeContext || undefined,
            rollingSummary,
            session.transcript.length + 1,
          );
          const messages = buildConversationMessages(session, trimmed);
          const lifelineQ = await generateValidQuestion(
            lifelinePrompt,
            [...messages, { role: 'user', content: 'I give up. I cannot figure this out.' }],
            getConversationModel(),
            session.intensity,
          );

          newLine();
          displayDim(`[Lifeline — one more question before you go]`);
          console.log(lifelineQ);
          newLine();
          currentQuestion = lifelineQ;
          continue;
        } else {
          addTurn(session, currentQuestion, '[gave up]', model);
          rl.close();
          return { outcome: 'abandoned' };
        }
      }

      // Normal response — process it
      addTurn(session, currentQuestion, trimmed, model);

      // Check for new file paths in response
      const newPaths = detectFilePaths(trimmed);
      for (const filePath of newPaths) {
        if (!codeFiles.has(filePath)) {
          const content = await readCodeFile(filePath);
          if (content) {
            codeFiles.set(filePath, content);
            addCodeFile(session, filePath);
          }
        }
      }

      // Check context window
      const contextResult = await checkAndSummarize(
        systemPrompt,
        session.transcript,
        rollingSummary,
      );
      rollingSummary = contextResult.summary;

      // Select contextual quote
      const quote = selectQuote(trimmed);
      const suggestedQuoteText = quote ? quote.text : undefined;

      // Build next turn
      const turnNumber = session.transcript.length + 1;
      const updatedCodeContext = formatCodeContext(codeFiles);
      systemPrompt = getSystemPrompt(
        session.intensity,
        session.problemDescription,
        updatedCodeContext || undefined,
        rollingSummary,
        turnNumber,
        suggestedQuoteText,
      );

      // Determine model — Sonnet for first analysis of new files
      model = newPaths.length > 0 ? getModelForFile(newPaths[0]) : getConversationModel();

      const messages = buildConversationMessages(session);
      const nextQuestion = await generateValidQuestion(
        systemPrompt,
        messages,
        model,
        session.intensity,
      );

      newLine();
      displayDim(`[Turn ${turnNumber}]`);
      console.log(nextQuestion);
      newLine();
      currentQuestion = nextQuestion;
    }
  } catch (error) {
    // Handle Ctrl+C / Ctrl+D / readline close
    if ((error as NodeJS.ErrnoException).code === 'ERR_USE_AFTER_CLOSE') {
      return { outcome: 'abandoned' };
    }
    rl.close();
    return { outcome: 'abandoned' };
  }
}

async function generateValidQuestion(
  systemPrompt: string,
  messages: Array<{ role: 'user' | 'assistant'; content: string }>,
  model: string,
  intensity: Session['intensity'],
): Promise<string> {
  let fullText = '';

  await streamQuestion(systemPrompt, messages, model, {
    onText: (text) => streamToTerminal(text),
    onComplete: (text) => {
      fullText = text;
    },
    onError: () => {},
  });

  // Clear the streamed output line
  process.stdout.write('\r\x1b[K');

  const validation = validateResponse(fullText);
  if (validation.valid) return fullText;

  // Try reprompt
  const reprompted = await repromptIfInvalid(systemPrompt, messages, fullText);
  if (reprompted) return reprompted;

  // Fallback
  return getFallbackQuestion(intensity);
}

function buildConversationMessages(
  session: Session,
  currentResponse?: string,
): Array<{ role: 'user' | 'assistant'; content: string }> {
  const messages: Array<{ role: 'user' | 'assistant'; content: string }> = [
    { role: 'user', content: session.problemDescription },
  ];

  for (const turn of session.transcript) {
    messages.push({ role: 'assistant', content: turn.question });
    if (turn.response && !turn.response.startsWith('[')) {
      messages.push({ role: 'user', content: turn.response });
    }
  }

  if (currentResponse) {
    messages.push({ role: 'user', content: currentResponse });
  }

  return messages;
}

File: nietzschian_debugger/core/session_loop.py
"""Interactive session loop for the debugging session."""

from __future__ import annotations

import sys
from dataclasses import dataclass
from typing import Optional

from ..types import Session, SessionOutcome
from ..llm.client import stream_question, get_conversation_model, get_model_for_file, StreamCallbacks
from ..llm.prompts import get_system_prompt
from ..llm.validator import validate_response, reprompt_if_invalid, get_fallback_question
from ..storage.file_reader import detect_file_paths, read_code_file, format_code_context
from .session import add_turn, add_code_file
from .context_manager import check_and_summarize, build_message_array
from ..quotes.selector import select_quote
from ..ui.renderer import (
    stream_to_terminal,
    new_line,
    display_session_header,
    display_dim,
)

EXIT_COMMANDS = {"exit", "quit"}
SOLVE_COMMANDS = {"solved", "found it"}


@dataclass
class SessionLoopResult:
    outcome: SessionOutcome


async def run_session_loop(session: Session) -> SessionLoopResult:
    """Run the interactive debugging session loop."""
    code_files: dict[str, str] = {}
    rolling_summary: Optional[str] = None
    lifeline_offered = False

    display_session_header(session.intensity)

    # Read initial code files from problem description
    initial_paths = detect_file_paths(session.problem_description)
    for file_path in initial_paths:
        content = await read_code_file(file_path)
        if content:
            code_files[file_path] = content
            add_code_file(session, file_path)

    # Generate opening question
    model = get_model_for_file(initial_paths[0]) if initial_paths else get_conversation_model()
    code_context = format_code_context(code_files)
    system_prompt = get_system_prompt(
        session.intensity,
        session.problem_description,
        code_context or None,
        rolling_summary,
        1,
    )

    opening_question = _generate_valid_question(
        system_prompt,
        [{"role": "user", "content": session.problem_description}],
        model,
        session.intensity,
    )

    new_line()
    display_dim("[Turn 1]")
    print(opening_question)
    new_line()

    current_question = opening_question

    try:
        while True:
            try:
                raw_input = input("> ")
            except EOFError:
                return SessionLoopResult(outcome="abandoned")

            trimmed = raw_input.strip()
            if not trimmed:
                continue

            lower = trimmed.lower()

            # Check exit commands
            if lower in EXIT_COMMANDS:
                add_turn(session, current_question, "[exited]", model)
                return SessionLoopResult(outcome="abandoned")

            # Check solve commands
            if lower in SOLVE_COMMANDS:
                add_turn(session, current_question, "[solved]", model)
                return SessionLoopResult(outcome="solved")

            # Check "I give up"
            if lower == "i give up":
                if not lifeline_offered:
                    lifeline_offered = True
                    add_turn(session, current_question, trimmed, model)

                    # Generate one lifeline question
                    lifeline_prompt = get_system_prompt(
                        session.intensity,
                        session.problem_description,
                        code_context or None,
                        rolling_summary,
                        len(session.transcript) + 1,
                    )
                    messages = _build_conversation_messages(session, trimmed)
                    lifeline_q = _generate_valid_question(
                        lifeline_prompt,
                        [*messages, {"role": "user", "content": "I give up. I cannot figure this out."}],
                        get_conversation_model(),
                        session.intensity,
                    )

                    new_line()
                    display_dim("[Lifeline -- one more question before you go]")
                    print(lifeline_q)
                    new_line()
                    current_question = lifeline_q
                    continue
                else:
                    add_turn(session, current_question, "[gave up]", model)
                    return SessionLoopResult(outcome="abandoned")

            # Normal response
            add_turn(session, current_question, trimmed, model)

            # Check for new file paths in response
            new_paths = detect_file_paths(trimmed)
            for file_path in new_paths:
                if file_path not in code_files:
                    content = await read_code_file(file_path)
                    if content:
                        code_files[file_path] = content
                        add_code_file(session, file_path)

            # Check context window
            context_turns, context_summary = await check_and_summarize(
                system_prompt,
                session.transcript,
                rolling_summary,
            )
            rolling_summary = context_summary

            # Select contextual quote
            quote = select_quote(trimmed)
            suggested_quote_text = quote.text if quote else None

            # Build next turn
            turn_number = len(session.transcript) + 1
            updated_code_context = format_code_context(code_files)
            system_prompt = get_system_prompt(
                session.intensity,
                session.problem_description,
                updated_code_context or None,
                rolling_summary,
                turn_number,
                suggested_quote_text,
            )

            # Determine model
            model = get_model_for_file(new_paths[0]) if new_paths else get_conversation_model()

            messages = _build_conversation_messages(session)
            next_question = _generate_valid_question(
                system_prompt,
                messages,
                model,
                session.intensity,
            )

            new_line()
            display_dim(f"[Turn {turn_number}]")
            print(next_question)
            new_line()
            current_question = next_question

    except KeyboardInterrupt:
        return SessionLoopResult(outcome="abandoned")
    except Exception:
        return SessionLoopResult(outcome="abandoned")


def _generate_valid_question(
    system_prompt: str,
    messages: list[dict[str, str]],
    model: str,
    intensity: str,
) -> str:
    """Generate a valid question, with reprompt and fallback."""
    full_text = ""

    def on_text(text: str) -> None:
        stream_to_terminal(text)

    def on_complete(text: str) -> None:
        nonlocal full_text
        full_text = text

    def on_error(error: Exception) -> None:
        pass

    stream_question(
        system_prompt,
        messages,
        model,
        StreamCallbacks(on_text=on_text, on_complete=on_complete, on_error=on_error),
    )

    # Clear the streamed output line
    sys.stdout.write("\r\033[K")

    validation = validate_response(full_text)
    if validation.valid:
        return full_text

    # Try reprompt
    reprompted = reprompt_if_invalid(system_prompt, messages, full_text)
    if reprompted:
        return reprompted

    # Fallback
    return get_fallback_question(intensity)  # type: ignore[arg-type]


def _build_conversation_messages(
    session: Session,
    current_response: Optional[str] = None,
) -> list[dict[str, str]]:
    """Build the conversation message history for the LLM."""
    messages: list[dict[str, str]] = [
        {"role": "user", "content": session.problem_description},
    ]

    for turn in session.transcript:
        messages.append({"role": "assistant", "content": turn.question})
        if turn.response and not turn.response.startswith("["):
            messages.append({"role": "user", "content": turn.response})

    if current_response:
        messages.append({"role": "user", "content": current_response})

    return messages
